packages/flutter/lib/src/services/hardware_keyboard.dart

```diff
@@ -618,25 +618,52 @@ enum KeyDataTransitMode {
   keyDataThenRawKeyData,
 }
 
-/// The assumbled information corresponding to a native key message.
+/// The assembled information converted from a native key message.
 ///
-/// While Flutter's [KeyEvent]s are created from key messages from the native
-/// platform, every native message might result in multiple [KeyEvent]s. For
-/// example, this might happen in order to synthesize missed modifier key
-/// presses or releases.
+/// Native key messages, meaning physically pressing or releasing
+/// keyboard keys, are translated into two different event streams in Flutter:
 ///
-/// A [KeyMessage] bundles all information related to a native key message
-/// together for the convenience of propagation on the [FocusNode] tree.
+///  * The [KeyEvent] stream, represented by [KeyMessage.events] (recommended).
+///  * The [RawKeyEvent] stream, represented by [KeyMessage.rawEvent] (legacy,
+///    to be deprecated).
 ///
-/// When dispatched to handlers or listeners, or propagated through the
-/// [FocusNode] tree, all handlers or listeners belonging to a node are
-/// executed regardless of their [KeyEventResult], and all results are combined
-/// into the result of the node using [combineKeyEventResults]. Empty [events]
-/// or [rawEvent] should be considered as a result of [KeyEventResult.ignored].
+/// Either the [KeyEvent] stream or the [RawKeyEvent] stream alone provides a
+/// complete description of the keyboard messages, but in different event
+/// models. Flutter is still transitioning from the legacy model to the new
+/// model, therefore dispatches both streams simultaneously before the
+/// transition is completed. [KeyMessage] is therefore used to bundles the
+/// stream segments of both models from a native key message together for the
+/// convenience of propagation.
 ///
-/// In very rare cases, a native key message might not result in a [KeyMessage].
-/// For example, key messages for Fn key are ignored on macOS for the
-/// convenience of cross-platform code.
+/// Typically, an application only needs to process either [KeyMessage.events]
+/// or [KeyMessage.rawEvent], not both. For example, when handling a
+/// [KeyMessage], simply handle each of [KeyMessage.events] independently.
+///
+/// In advanced cases, a widget needs to process both streams at the same time.
+/// For example, [FocusNode] has an `onKey` that dispatches [RawKeyEvent]s and
+/// an `onKeyEvent` that dispatches [KeyEvent]s. To processes a [KeyMessage],
+/// it first calls `onKeyEvent` with each [KeyEvent] of [events], and then
+/// `onKey` with [rawEvent]. All callbacks are invoked regardless of their
+/// [KeyEventResult]. Their results are combined into the result of the node
+/// using [combineKeyEventResults].
+///
+/// {@tool snippet}
+/// ```dart
```

Comment:
You might add some text here to introduce the snippet, since it'll be inside a blue box.

You could just move the previous paragraph inside of the `{@tool snippet}` if you prefer.